File: library/src/main/java/com/yy/codex/uikit/UIViewAnimator.java
package com.yy.codex.uikit;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.facebook.rebound.Spring;
import com.facebook.rebound.SpringConfig;
import com.facebook.rebound.SpringListener;
import com.facebook.rebound.SpringSystem;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by cuiminghui on 2017/1/16.
 */

public class UIViewAnimator {

    @Nullable
    static private HashMap<UIView, HashMap<String, UIViewAnimatorState>> sAnimationState = null;

    static void addAnimationState(@NonNull UIView view, @NonNull String aKey, double originValue, double finalValue) {
        if (sAnimationState == null) {
            return;
        }
        if (originValue == finalValue) {
            return;
        }
        if (sAnimationState.get(view) == null) {
            sAnimationState.put(view, new HashMap<String, UIViewAnimatorState>());
        }
        UIViewAnimatorState<Number> log = new UIViewAnimatorState<>();
        log.valueType = 1;
        log.originValue = originValue;
        log.finalValue = finalValue;
        sAnimationState.get(view).put(aKey, log);
    }

    static private void resetAnimationState() {
        sAnimationState = new HashMap<>();
    }

    static public UIViewAnimation linear(@NonNull Runnable animations) {
        return linear(0.25, animations, null);
    }

    static public UIViewAnimation linear(double duration, @NonNull Runnable animations, @Nullable final Runnable completion) {
        final UIViewAnimation animation = new UIViewAnimation();
        resetAnimationState();
        animations.run();
        final int[] aniCount = {0};
        for (final Map.Entry<UIView, HashMap<String, UIViewAnimatorState>> viewProps: sAnimationState.entrySet()) {
            for (final Map.Entry<String, UIViewAnimatorState> animateProp: viewProps.getValue().entrySet()) {
                UIViewAnimatorState log = animateProp.getValue();
                if (log.valueType == 1) {
                    aniCount[0]++;
                    viewProps.getKey().animate(animateProp.getKey(), (float)((double)log.originValue));
                    ValueAnimator animator = ValueAnimator.ofFloat((float)((double)log.originValue), (float)((double)log.finalValue));
                    animator.setDuration((long) (duration * 1000));
                    animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                        @Override
                        public void onAnimationUpdate(@NonNull ValueAnimator valueAnimator) {
                            if (animation.isCancelled()) {
                                return;
                            }
                            float currentValue = (float)valueAnimator.getAnimatedValue();
                            viewProps.getKey().animate(animateProp.getKey(), currentValue);
                        }
                    });
                    animator.addListener(new Animator.AnimatorListener() {
                        @Override
                        public void onAnimationStart(Animator animator) {

                        }

                        @Override
                        public void onAnimationEnd(Animator animator) {
                            if (animation.isCancelled()) {
                                return;
                            }
                            aniCount[0]--;
                            if (aniCount[0] <= 0) {
                                animation.markFinished();
                            }
                            if (aniCount[0] <= 0 && completion != null) {
                                completion.run();
                            }
                        }

                        @Override
                        public void onAnimationCancel(Animator animator) {

                        }

                        @Override
                        public void onAnimationRepeat(Animator animator) {

                        }
                    });
                    animator.setTarget(viewProps.getKey());
                    animator.start();
                }
            }
        }
        if (aniCount[0] <= 0) {
            animation.markFinished();
            if (completion != null) {
                completion.run();
            }
        }
        sAnimationState = null;
        return animation;
    }

    static public UIViewAnimation spring(@NonNull Runnable animations) {
        return spring(animations, null);
    }

    static public UIViewAnimation spring(@NonNull Runnable animations, @Nullable final Runnable completion) {
        final UIViewAnimation animation = new UIViewAnimation();
        resetAnimationState();
        animations.run();
        final int[] aniCount = {0};
        SpringSystem system = SpringSystem.create();
        for (final Map.Entry<UIView, HashMap<String, UIViewAnimatorState>> viewProps: sAnimationState.entrySet()) {
            for (final Map.Entry<String, UIViewAnimatorState> animateProp: viewProps.getValue().entrySet()) {
                final UIViewAnimatorState log = animateProp.getValue();
                if (log.valueType == 1) {
                    aniCount[0]++;
                    viewProps.getKey().animate(animateProp.getKey(), (float)((double)log.originValue));
                    Spring spring = system.createSpring();
                    spring.setCurrentValue((float)((double)log.originValue));
                    spring.addListener(new SpringListener() {
                        @Override
                        public void onSpringUpdate(@NonNull Spring spring) {
                            if (animation.isCancelled()) {
                                return;
                            }
                            float currentValue = (float)spring.getCurrentValue();
                            viewProps.getKey().animate(animateProp.getKey(), currentValue);
                        }

                        @Override
                        public void onSpringAtRest(@NonNull Spring spring) {
                            if (animation.isCancelled()) {
                                return;
                            }
                            float currentValue = (float)spring.getCurrentValue();
                            viewProps.getKey().animate(animateProp.getKey(), currentValue);
                            aniCount[0]--;
                            if (aniCount[0] <= 0 && completion != null) {
                                completion.run();
                            }
                        }
                        @Override
                        public void onSpringActivate(Spring spring) {
                        }

                        @Override
                        public void onSpringEndStateChange(Spring spring) {
                        }
                    });
                    spring.setEndValue((float)((double)log.finalValue));
                }
            }
        }
        if (aniCount[0] <= 0) {
            animation.markFinished();
            if (completion != null) {
                completion.run();
            }
        }
        sAnimationState = null;
        return animation;
    }

    static public UIViewAnimation springWithOptions(double tension, double friction, double velocity, @NonNull Runnable animations, @Nullable final Runnable completion) {
        final UIViewAnimation animation = new UIViewAnimation();
        resetAnimationState();
        animations.run();
        final int[] aniCount = {0};
        SpringSystem system = SpringSystem.create();
        for (final Map.Entry<UIView, HashMap<String, UIViewAnimatorState>> viewProps: sAnimationState.entrySet()) {
            for (final Map.Entry<String, UIViewAnimatorState> animateProp: viewProps.getValue().entrySet()) {
                UIViewAnimatorState log = animateProp.getValue();
                if (log.valueType == 1) {
                    aniCount[0]++;
                    viewProps.getKey().animate(animateProp.getKey(), (float)((double)log.originValue));
                    Spring spring = system.createSpring();
                    spring.setCurrentValue((float)((double)log.originValue));
                    SpringConfig config = new SpringConfig(tension, friction);
                    spring.setSpringConfig(config);
                    spring.setVelocity(velocity);
                    spring.addListener(new SpringListener() {
                        @Override
                        public void onSpringUpdate(@NonNull Spring spring) {
                            if (animation.isCancelled()) {
                                return;
                            }
                            float currentValue = (float)spring.getCurrentValue();
                            viewProps.getKey().animate(animateProp.getKey(), currentValue);
                        }

                        @Override
                        public void onSpringAtRest(@NonNull Spring spring) {
                            if (animation.isCancelled()) {
                                return;
                            }
                            float currentValue = (float)spring.getCurrentValue();
                            viewProps.getKey().animate(animateProp.getKey(), currentValue);
                            aniCount[0]--;
                            if (aniCount[0] <= 0) {
                                animation.markFinished();
                            }
                            if (aniCount[0] <= 0 && completion != null) {
                                completion.run();
                            }
                        }

                        @Override
                        public void onSpringActivate(Spring spring) {
                        }

                        @Override
                        public void onSpringEndStateChange(Spring spring) {
                        }
                    });
                    spring.setEndValue((float)((double)log.finalValue));
                }
            }
        }
        if (aniCount[0] <= 0) {
            animation.markFinished();
            if (completion != null) {
                completion.run();
            }
        }
        sAnimationState = null;
        return animation;
    }

    static public double decayDeceleration = 0.997;

    static public UIViewAnimation decay(final UIView animationView, final String animationKey, final double fromValue, final double velocity, @Nullable final Runnable completion) {
        final UIViewAnimation animation = new UIViewAnimation();
        final long startTime = System.currentTimeMillis();
        final double deceleration = decayDeceleration;
        final double finalValue = fromValue + (velocity / (1.0 - deceleration)) * (1 - Math.exp(-(1 - deceleration) * (999999999)));
        ValueAnimator valueAnimator = ValueAnimator.ofInt(0, 1);
        valueAnimator.setDuration(16);
        valueAnimator.setRepeatCount(9999999);
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                if (animation.isCancelled()) {
                    valueAnimator.cancel();
                    return;
                }
                long now = System.currentTimeMillis();
                double value = fromValue + (velocity / (1.0 - deceleration)) * (1 - Math.exp(-(1 - deceleration) * (now - startTime)));
                animationView.animate(animationKey, (float) value);
                if (Math.abs(finalValue - value) < 0.1) {
                    valueAnimator.cancel();
                    animation.markFinished();
                    if (completion != null) {
                        completion.run();
                    }
                }
            }
        });
        valueAnimator.start();
        return animation;
    }

    static public UIViewAnimation decayToValue(final UIView animationView, final String animationKey, final double fromValue, final double toValue, @Nullable final Runnable completion) {
        final double deceleration = decayDeceleration;
        double velocity = (toValue / (1 - Math.exp(-(1 - deceleration) * (999999999))) - fromValue) * (1.0 - deceleration);
        return decay(animationView, animationKey, fromValue, velocity, completion);
    }

    static public class UIViewAnimationDecayBoundsOptions {
        double fromValue;
        double velocity;
        boolean allowBounds = true;
        boolean alwaysBounds = false;
        double topBounds;
        double bottomBounds;
        double viewBounds;
    }

    static public UIViewAnimation decayBounds(final UIView animationView, final String animationKey, final UIViewAnimationDecayBoundsOptions options, @Nullable final Runnable completion) {
        if (options.bottomBounds - options.topBounds < options.viewBounds) {
            if (options.allowBounds && options.alwaysBounds) {
                options.bottomBounds = options.topBounds;
            }
            else {
                return new UIViewAnimation();
            }
        }
        final double deceleration = decayDeceleration;
        final double finalValue = options.fromValue + (options.velocity / (1.0 - deceleration)) * (1 - Math.exp(-(1 - deceleration) * (999999999)));
        double backStartValue;
        final boolean[] backStarted = new boolean[1];
        final double backEndValue;
        if (finalValue < options.topBounds) {
            double tmpFinalValue = finalValue;
            if (options.fromValue < options.topBounds) {
                tmpFinalValue = options.topBounds + (options.velocity / (1.0 - deceleration)) * (1 - Math.exp(-(1 - deceleration) * (999999999)));
            }
            backStartValue = options.topBounds - (options.topBounds - tmpFinalValue) / 12.0;
            if (options.fromValue < options.topBounds) {
                backStartValue += (options.fromValue - options.topBounds);
            }
            backEndValue = options.topBounds;
            if (options.fromValue < options.topBounds && Math.abs(options.velocity) < 1.0) {
                backStarted[0] = true;
            }
        }
        else if (finalValue > options.bottomBounds) {
            double tmpFinalValue = finalValue;
            if (options.fromValue > options.bottomBounds) {
                tmpFinalValue = options.topBounds + (options.velocity / (1.0 - deceleration)) * (1 - Math.exp(-(1 - deceleration) * (999999999)));
            }
            backStartValue = (finalValue - options.bottomBounds) / 12.0 + options.bottomBounds;
            if (options.fromValue > options.bottomBounds) {
                backStartValue += (options.fromValue - options.bottomBounds);
            }
            backEndValue = options.bottomBounds;
            if (options.fromValue > options.bottomBounds && Math.abs(options.velocity) < 1.0) {
                backStarted[0] = true;
            }
        }
        else {
            return decay(animationView, animationKey, options.fromValue, options.velocity, completion);
        }
        final long startTime = System.currentTimeMillis();
        final UIViewAnimation animation = new UIViewAnimation();
        ValueAnimator valueAnimator = ValueAnimator.ofInt(0, 1);
        valueAnimator.setDuration(16);
        valueAnimator.setRepeatCount(9999999);
        final double finalBackStartValue = backStartValue;
        final double finalBackEndValue = backEndValue;
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                if (animation.isCancelled()) {
                    valueAnimator.cancel();
                    return;
                }
                double decayValue = options.fromValue + (options.velocity / (1.0 - deceleration)) * (1 - Math.exp(-(1 - deceleration) * (System.currentTimeMillis() - startTime)));
                if (options.fromValue < options.topBounds) {
                    decayValue = options.topBounds + (options.velocity / (1.0 - deceleration)) * (1 - Math.exp(-(1 - deceleration) * (System.currentTimeMillis() - startTime)));
                }
                if (options.fromValue > options.bottomBounds) {
                    decayValue = options.bottomBounds + (options.velocity / (1.0 - deceleration)) * (1 - Math.exp(-(1 - deceleration) * (System.currentTimeMillis() - startTime)));
                }
                if (decayValue < options.topBounds) {
                    decayValue = options.topBounds - (options.topBounds - decayValue) / 3;
                    if (options.fromValue < options.topBounds) {
                        decayValue += (options.fromValue - options.topBounds);
                    }
                }
                else if (decayValue > options.bottomBounds) {
                    decayValue = (decayValue - options.bottomBounds) / 3 + options.bottomBounds;
                    if (options.fromValue < options.topBounds) {
                        decayValue += (options.fromValue - options.bottomBounds);
                    }
                }
                if (backStarted[0]) {
                    SpringSystem system = SpringSystem.create();
                    Spring spring = system.createSpring();
                    spring.setCurrentValue(finalBackStartValue);
                    SpringConfig config = new SpringConfig(120.0, 20.0);
                    spring.setSpringConfig(config);
                    spring.addListener(new SpringListener() {
                        @Override
                        public void onSpringUpdate(Spring spring) {
                            if (animation.isCancelled()) {
                                return;
                            }
                            float currentValue = (float)spring.getCurrentValue();
                            animationView.animate(animationKey, currentValue);
                        }

                        @Override
                        public void onSpringAtRest(Spring spring) {
                            if (animation.isCancelled()) {
                                return;
                            }
                            if (completion != null) {
                                completion.run();
                            }
                        }

                        @Override
                        public void onSpringActivate(Spring spring) {

                        }

                        @Override
                        public void onSpringEndStateChange(Spring spring) {

                        }
                    });
                    spring.setEndValue(finalBackEndValue);
                    valueAnimator.cancel();
                }
                else if (!backStarted[0]) {
                    if (finalValue < options.topBounds && decayValue < finalBackStartValue && !backStarted[0]) {
                        backStarted[0] = true;
                    }
                    else if (finalValue > options.bottomBounds && decayValue > finalBackStartValue && !backStarted[0]) {
                        backStarted[0] = true;
                    }
                }
                else if (!options.allowBounds) {
                    if (decayValue < options.topBounds) {
                        valueAnimator.cancel();
                    }
                    else if (decayValue > options.bottomBounds) {
                        valueAnimator.cancel();
                    }
                }
                animationView.animate(animationKey, (float) decayValue);
            }
        });
        valueAnimator.start();
        return animation;
    }

}

File: library/src/main/java/com/yy/codex/uikit/UIViewAnimator.kt
package com.yy.codex.uikit

import android.animation.Animator
import android.animation.ValueAnimator

import com.facebook.rebound.Spring
import com.facebook.rebound.SpringConfig
import com.facebook.rebound.SpringListener
import com.facebook.rebound.SpringSystem

import java.util.HashMap

/**
 * Created by cuiminghui on 2017/1/16.
 */

object UIViewAnimator {

    private var sAnimationState: HashMap<UIView, HashMap<String, UIViewAnimatorState<*>>> = hashMapOf()

    fun addAnimationState(view: UIView, aKey: String, originValue: Double, finalValue: Double) {
        if (sAnimationState == null) {
            return
        }
        if (originValue == finalValue) {
            return
        }
        if (sAnimationState[view] == null) {
            sAnimationState.put(view, HashMap<String, UIViewAnimatorState<*>>())
        }
        val log = UIViewAnimatorState<Number>()
        log.valueType = 1
        log.originValue = originValue
        log.finalValue = finalValue
        sAnimationState[view]?.let {
            it.put(aKey, log)
        }
    }

    private fun resetAnimationState() {
        sAnimationState = hashMapOf()
    }

    fun linear(animations: Runnable): UIViewAnimation {
        return linear(0.25, animations, null)
    }

    fun linear(duration: Double, animations: Runnable, completion: Runnable?): UIViewAnimation {
        val animation = UIViewAnimation()
        resetAnimationState()
        animations.run()
        val aniCount = intArrayOf(0)
        for ((key, value) in sAnimationState) {
            for ((key1, log) in value) {
                if (log.valueType == 1) {
                    aniCount[0]++
                    key.animate(key1, (log.originValue as Double).toFloat())
                    val animator = ValueAnimator.ofFloat((log.originValue as Double).toFloat(), (log.finalValue as Double).toFloat())
                    animator.duration = (duration * 1000).toLong()
                    animator.addUpdateListener(ValueAnimator.AnimatorUpdateListener { valueAnimator ->
                        if (animation.isCancelled) {
                            return@AnimatorUpdateListener
                        }
                        val currentValue = valueAnimator.animatedValue as Float
                        key.animate(key1, currentValue)
                    })
                    animator.addListener(object : Animator.AnimatorListener {
                        override fun onAnimationStart(animator: Animator) {

                        }

                        override fun onAnimationEnd(animator: Animator) {
                            if (animation.isCancelled) {
                                return
                            }
                            aniCount[0]--
                            if (aniCount[0] <= 0) {
                                animation.markFinished()
                            }
                            if (aniCount[0] <= 0 && completion != null) {
                                completion.run()
                            }
                        }

                        override fun onAnimationCancel(animator: Animator) {

                        }

                        override fun onAnimationRepeat(animator: Animator) {

                        }
                    })
                    animator.setTarget(key)
                    animator.start()
                }
            }
        }
        if (aniCount[0] <= 0) {
            animation.markFinished()
            completion?.run()
        }
        sAnimationState = hashMapOf()
        return animation
    }

    @JvmOverloads fun spring(animations: Runnable, completion: Runnable? = null): UIViewAnimation {
        val animation = UIViewAnimation()
        resetAnimationState()
        animations.run()
        val aniCount = intArrayOf(0)
        val system = SpringSystem.create()
        for ((key, value) in sAnimationState) {
            for ((key1, log) in value) {
                if (log.valueType == 1) {
                    aniCount[0]++
                    key.animate(key1, (log.originValue as Double).toFloat())
                    val spring = system.createSpring()
                    spring.currentValue = (log.originValue as Double).toFloat().toDouble()
                    spring.addListener(object : SpringListener {
                        override fun onSpringUpdate(spring: Spring) {
                            if (animation.isCancelled) {
                                return
                            }
                            val currentValue = spring.currentValue.toFloat()
                            key.animate(key1, currentValue)
                        }

                        override fun onSpringAtRest(spring: Spring) {
                            if (animation.isCancelled) {
                                return
                            }
                            val currentValue = spring.currentValue.toFloat()
                            key.animate(key1, currentValue)
                            aniCount[0]--
                            if (aniCount[0] <= 0 && completion != null) {
                                completion.run()
                            }
                        }

                        override fun onSpringActivate(spring: Spring) {}

                        override fun onSpringEndStateChange(spring: Spring) {}
                    })
                    spring.endValue = (log.finalValue as Double).toFloat().toDouble()
                }
            }
        }
        if (aniCount[0] <= 0) {
            animation.markFinished()
            completion?.run()
        }
        sAnimationState = hashMapOf()
        return animation
    }

    fun springWithOptions(tension: Double, friction: Double, velocity: Double, animations: Runnable, completion: Runnable?): UIViewAnimation {
        val animation = UIViewAnimation()
        resetAnimationState()
        animations.run()
        val aniCount = intArrayOf(0)
        val system = SpringSystem.create()
        for ((key, value) in sAnimationState) {
            for ((key1, log) in value) {
                if (log.valueType == 1) {
                    aniCount[0]++
                    key.animate(key1, (log.originValue as Double).toFloat())
                    val spring = system.createSpring()
                    spring.currentValue = (log.originValue as Double).toFloat().toDouble()
                    val config = SpringConfig(tension, friction)
                    spring.springConfig = config
                    spring.velocity = velocity
                    spring.addListener(object : SpringListener {
                        override fun onSpringUpdate(spring: Spring) {
                            if (animation.isCancelled) {
                                return
                            }
                            val currentValue = spring.currentValue.toFloat()
                            key.animate(key1, currentValue)
                        }

                        override fun onSpringAtRest(spring: Spring) {
                            if (animation.isCancelled) {
                                return
                            }
                            val currentValue = spring.currentValue.toFloat()
                            key.animate(key1, currentValue)
                            aniCount[0]--
                            if (aniCount[0] <= 0) {
                                animation.markFinished()
                            }
                            if (aniCount[0] <= 0 && completion != null) {
                                completion.run()
                            }
                        }

                        override fun onSpringActivate(spring: Spring) {}

                        override fun onSpringEndStateChange(spring: Spring) {}
                    })
                    spring.endValue = (log.finalValue as Double).toFloat().toDouble()
                }
            }
        }
        if (aniCount[0] <= 0) {
            animation.markFinished()
            completion?.run()
        }
        sAnimationState = hashMapOf()
        return animation
    }

    var decayDeceleration = 0.997

    fun decay(animationView: UIView, animationKey: String, fromValue: Double, velocity: Double, completion: Runnable?): UIViewAnimation {
        val animation = UIViewAnimation()
        val startTime = System.currentTimeMillis()
        val deceleration = decayDeceleration
        val finalValue = fromValue + velocity / (1.0 - deceleration) * (1 - Math.exp(-(1 - deceleration) * 999999999))
        val valueAnimator = ValueAnimator.ofInt(0, 1)
        valueAnimator.duration = 16
        valueAnimator.repeatCount = 9999999
        valueAnimator.addUpdateListener(ValueAnimator.AnimatorUpdateListener { valueAnimator ->
            if (animation.isCancelled) {
                valueAnimator.cancel()
                return@AnimatorUpdateListener
            }
            val now = System.currentTimeMillis()
            val value = fromValue + velocity / (1.0 - deceleration) * (1 - Math.exp(-(1 - deceleration) * (now - startTime)))
            animationView.animate(animationKey, value.toFloat())
            if (Math.abs(finalValue - value) < 0.1) {
                valueAnimator.cancel()
                animation.markFinished()
                completion?.run()
            }
        })
        valueAnimator.start()
        return animation
    }

    fun decayToValue(animationView: UIView, animationKey: String, fromValue: Double, toValue: Double, completion: Runnable?): UIViewAnimation {
        val deceleration = decayDeceleration
        val velocity = (toValue / (1 - Math.exp(-(1 - deceleration) * 999999999)) - fromValue) * (1.0 - deceleration)
        return decay(animationView, animationKey, fromValue, velocity, completion)
    }

    class UIViewAnimationDecayBoundsOptions {
        var fromValue: Double = 0.toDouble()
        var velocity: Double = 0.toDouble()
        var allowBounds = true
        var alwaysBounds = false
        var topBounds: Double = 0.toDouble()
        var bottomBounds: Double = 0.toDouble()
        var viewBounds: Double = 0.toDouble()
    }

    fun decayBounds(animationView: UIView, animationKey: String, options: UIViewAnimationDecayBoundsOptions, completion: Runnable?): UIViewAnimation {
        if (!options.allowBounds && (options.fromValue <= options.topBounds || options.fromValue >= options.bottomBounds)) {
            return UIViewAnimation()
        }
        if (options.bottomBounds - options.topBounds < options.viewBounds) {
            if (options.allowBounds && options.alwaysBounds) {
                options.bottomBounds = options.topBounds
            } else {
                return UIViewAnimation()
            }
        }
        val deceleration = decayDeceleration
        val finalValue = options.fromValue + options.velocity / (1.0 - deceleration) * (1 - Math.exp(-(1 - deceleration) * 999999999))
        var backStartValue: Double
        val backStarted = BooleanArray(1)
        val backEndValue: Double
        if (finalValue < options.topBounds) {
            var tmpFinalValue = finalValue
            if (options.fromValue < options.topBounds) {
                tmpFinalValue = options.topBounds + options.velocity / (1.0 - deceleration) * (1 - Math.exp(-(1 - deceleration) * 999999999))
            }
            backStartValue = options.topBounds - (options.topBounds - tmpFinalValue) / 12.0
            if (options.fromValue < options.topBounds) {
                backStartValue += options.fromValue - options.topBounds
            }
            backEndValue = options.topBounds
            if (options.fromValue < options.topBounds && Math.abs(options.velocity) < 1.0) {
                backStarted[0] = true
            }
        } else if (finalValue > options.bottomBounds) {
            var tmpFinalValue = finalValue
            if (options.fromValue > options.bottomBounds) {
                tmpFinalValue = options.bottomBounds + options.velocity / (1.0 - deceleration) * (1 - Math.exp(-(1 - deceleration) * 999999999))
            }
            backStartValue = (tmpFinalValue - options.bottomBounds) / 12.0 + options.bottomBounds
            if (options.fromValue > options.bottomBounds) {
                backStartValue += options.fromValue - options.bottomBounds
            }
            backEndValue = options.bottomBounds
            if (options.fromValue > options.bottomBounds && Math.abs(options.velocity) < 1.0) {
                backStarted[0] = true
            }
        } else {
            return decay(animationView, animationKey, options.fromValue, options.velocity, completion)
        }
        val startTime = System.currentTimeMillis()
        val animation = UIViewAnimation()
        val valueAnimator = ValueAnimator.ofInt(0, 1)
        valueAnimator.duration = 16
        valueAnimator.repeatCount = 9999999
        val finalBackStartValue = backStartValue
        val finalBackEndValue = backEndValue
        valueAnimator.addUpdateListener(ValueAnimator.AnimatorUpdateListener { valueAnimator ->
            if (animation.isCancelled) {
                valueAnimator.cancel()
                return@AnimatorUpdateListener
            }
            var decayValue = options.fromValue + options.velocity / (1.0 - deceleration) * (1 - Math.exp(-(1 - deceleration) * (System.currentTimeMillis() - startTime)))
            if (options.fromValue < options.topBounds) {
                decayValue = options.topBounds + options.velocity / (1.0 - deceleration) * (1 - Math.exp(-(1 - deceleration) * (System.currentTimeMillis() - startTime)))
            }
            if (options.fromValue > options.bottomBounds) {
                decayValue = options.bottomBounds + options.velocity / (1.0 - deceleration) * (1 - Math.exp(-(1 - deceleration) * (System.currentTimeMillis() - startTime)))
            }
            if (decayValue < options.topBounds) {
                decayValue = options.topBounds - (options.topBounds - decayValue) / 3
                if (options.fromValue < options.topBounds) {
                    decayValue -= options.topBounds - options.fromValue
                }
            } else if (decayValue > options.bottomBounds) {
                decayValue = (decayValue - options.bottomBounds) / 3 + options.bottomBounds
                if (options.fromValue > options.bottomBounds) {
                    decayValue += options.fromValue - options.bottomBounds
                }
            }
            if (!options.allowBounds) {
                if (decayValue <= options.topBounds) {
                    animationView.animate(animationKey, options.topBounds.toFloat())
                    valueAnimator.cancel()
                    animation.markFinished()
                    return@AnimatorUpdateListener
                } else if (decayValue >= options.bottomBounds) {
                    animationView.animate(animationKey, options.bottomBounds.toFloat())
                    valueAnimator.cancel()
                    animation.markFinished()
                    return@AnimatorUpdateListener
                }
            } else if (backStarted[0]) {
                val system = SpringSystem.create()
                val spring = system.createSpring()
                spring.currentValue = finalBackStartValue
                val config = SpringConfig(120.0, 20.0)
                spring.springConfig = config
                spring.addListener(object : SpringListener {
                    override fun onSpringUpdate(spring: Spring) {
                        if (animation.isCancelled) {
                            return
                        }
                        val currentValue = spring.currentValue.toFloat()
                        animationView.animate(animationKey, currentValue)
                    }

                    override fun onSpringAtRest(spring: Spring) {
                        if (animation.isCancelled) {
                            return
                        }
                        completion?.run()
                    }

                    override fun onSpringActivate(spring: Spring) {

                    }

                    override fun onSpringEndStateChange(spring: Spring) {

                    }
                })
                spring.endValue = finalBackEndValue
                valueAnimator.cancel()
            } else if (!backStarted[0]) {
                if (finalValue < options.topBounds && decayValue < finalBackStartValue && !backStarted[0]) {
                    backStarted[0] = true
                } else if (finalValue > options.bottomBounds && decayValue > finalBackStartValue && !backStarted[0]) {
                    backStarted[0] = true
                }
            }
            animationView.animate(animationKey, decayValue.toFloat())
        })
        valueAnimator.start()
        return animation
    }

}
